File: 2020/155.min-stack.js
/*
 * @lc app=leetcode id=155 lang=javascript
 *
 * [155] Min Stack
 */

// @lc code=start
/**
 * initialize your data structure here.
 */
var MinStack = function() {
  this.vals = []
  this.min = null
};

/** 
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function(x) {
  this.min = this.min == null ? x : Math.min(this.min, x)
  this.vals.push(x)
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
  let popped = this.vals.pop()
  if (this.vals.length == 0) this.min = null
  else if (this.min == popped) this.min = Math.min(...this.vals)
  return popped
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
  return this.vals[this.vals.length-1]
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
  return this.min
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */
// @lc code=end

File: 155.min-stack.py
#
# @lc app=leetcode id=155 lang=python3
#
# [155] Min Stack
#

# @lc code=start
class MinStack:

    def __init__(self):
        self.stack = []
        self.lowStack = []
        self.lowest = float('inf')

    def push(self, val: int) -> None:
        self.stack += [val]
        if val <= self.lowest:
            self.lowStack += [val]
            self.lowest = val

    def pop(self) -> None:
        if self.stack[-1] == self.lowStack[-1]:
            del self.lowStack[-1]
            self.lowest = self.lowStack[-1] if len(self.lowStack) else float('inf')
        del self.stack[-1]

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.lowest



# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
# @lc code=end

